File: internal/awiki/wiki/analysis.go
package wiki

import (
	"fmt"
	"math/rand"
	"sort"
	"strings"
)

type ShortestPathSample struct {
	Nodes  []string
	Length int
}

type AvgShortestPathReport struct {
	ComponentSize int
	SampleCount   int
	Average       float64
	LongerPaths   []ShortestPathSample
}

func (v *Vault) ApproximateAverageShortestPath(sampleCount, exampleCount int, seed int64) (AvgShortestPathReport, error) {
	component := v.largestConnectedComponentKeys()
	if len(component) < 2 {
		return AvgShortestPathReport{}, fmt.Errorf("largest connected component must contain at least two documents")
	}
	if sampleCount <= 0 {
		return AvgShortestPathReport{}, fmt.Errorf("sample count must be positive")
	}

	pairs := sampleComponentPairs(len(component), sampleCount, seed)
	samples := make([]ShortestPathSample, 0, len(pairs))
	totalLength := 0

	for _, pair := range pairs {
		path, err := v.shortestPathBetweenKeys(component[pair[0]], component[pair[1]])
		if err != nil {
			return AvgShortestPathReport{}, err
		}

		length := len(path) - 1
		totalLength += length
		samples = append(samples, ShortestPathSample{
			Nodes:  path,
			Length: length,
		})
	}

	report := AvgShortestPathReport{
		ComponentSize: len(component),
		SampleCount:   len(samples),
		Average:       float64(totalLength) / float64(len(samples)),
	}
	if exampleCount <= 0 {
		return report, nil
	}

	for _, sample := range samples {
		if float64(sample.Length) > report.Average {
			report.LongerPaths = append(report.LongerPaths, sample)
		}
	}
	sort.Slice(report.LongerPaths, func(i, j int) bool {
		if report.LongerPaths[i].Length != report.LongerPaths[j].Length {
			return report.LongerPaths[i].Length > report.LongerPaths[j].Length
		}
		return strings.ToLower(strings.Join(report.LongerPaths[i].Nodes, "\x00")) < strings.ToLower(strings.Join(report.LongerPaths[j].Nodes, "\x00"))
	})
	if len(report.LongerPaths) > exampleCount {
		report.LongerPaths = report.LongerPaths[:exampleCount]
	}

	return report, nil
}

func sampleComponentPairs(nodeCount, sampleCount int, seed int64) [][2]int {
	totalPairs := nodeCount * (nodeCount - 1) / 2
	if sampleCount >= totalPairs {
		pairs := make([][2]int, 0, totalPairs)
		for i := 0; i < nodeCount; i++ {
			for j := i + 1; j < nodeCount; j++ {
				pairs = append(pairs, [2]int{i, j})
			}
		}
		return pairs
	}

	rng := rand.New(rand.NewSource(seed))
	seen := make(map[uint64]struct{}, sampleCount)
	pairs := make([][2]int, 0, sampleCount)
	for len(pairs) < sampleCount {
		i := rng.Intn(nodeCount)
		j := rng.Intn(nodeCount - 1)
		if j >= i {
			j++
		}
		if i > j {
			i, j = j, i
		}

		key := uint64(i)<<32 | uint64(j)
		if _, ok := seen[key]; ok {
			continue
		}
		seen[key] = struct{}{}
		pairs = append(pairs, [2]int{i, j})
	}
	return pairs
}

File: src/wiki/analysis.rs
use rand::{rngs::StdRng, Rng, SeedableRng};
use rayon::prelude::*;
use rustc_hash::{FxHashMap, FxHashSet};
use std::collections::VecDeque;

use super::{AvgPathReport, LintReport, Vault, WantedPage, WantedSource};
use crate::{text::lower, wiki::path::normalize_document_name};

impl Vault {
    pub(crate) fn lint(&self) -> LintReport {
        let mut report = LintReport {
            document_count: self.documents.len(),
            largest_component_size: 0,
            covered_documents: 0,
            orphans: Vec::new(),
            islands: Vec::new(),
            link_only_lines: Vec::new(),
        };
        let mut visited = vec![false; self.documents.len()];
        let mut components = Vec::new();
        for (idx, doc) in self.documents.iter().enumerate() {
            if !doc.excerpt.trim().is_empty() {
                report.covered_documents += 1;
            }
            for issue in &doc.link_only {
                report.link_only_lines.push((idx, issue.clone()));
            }
            if self.undirected[idx].is_empty() {
                report.orphans.push(idx);
                visited[idx] = true;
                continue;
            }
            if visited[idx] {
                continue;
            }
            let mut component = self.collect_component(idx, &mut visited);
            self.sort_doc_indices(&mut component);
            components.push(component);
        }
        self.sort_doc_indices(&mut report.orphans);
        report.link_only_lines.sort_by(|(a, ia), (b, ib)| {
            lower(&self.documents[*a].name)
                .cmp(&lower(&self.documents[*b].name))
                .then(ia.line.cmp(&ib.line))
        });
        components.sort_by(|a, b| {
            b.len().cmp(&a.len()).then_with(|| {
                lower(&self.documents[a[0]].name).cmp(&lower(&self.documents[b[0]].name))
            })
        });
        if components.len() > 1 {
            report.islands = components[1..].to_vec();
        }
        if let Some(first) = components.first() {
            report.largest_component_size = first.len();
        } else if !report.orphans.is_empty() {
            report.largest_component_size = 1;
        }
        report
    }

    fn collect_component(&self, start: usize, visited: &mut [bool]) -> Vec<usize> {
        let mut q = VecDeque::from([start]);
        visited[start] = true;
        let mut component = Vec::new();
        while let Some(cur) = q.pop_front() {
            component.push(cur);
            let mut neighbors: Vec<_> = self.undirected[cur].iter().copied().collect();
            self.sort_doc_indices(&mut neighbors);
            for next in neighbors {
                if !visited[next] {
                    visited[next] = true;
                    q.push_back(next);
                }
            }
        }
        component
    }

    pub(crate) fn shortest_path_indices(
        &self,
        from: usize,
        to: usize,
    ) -> Result<Vec<usize>, String> {
        if from == to {
            return Ok(vec![from]);
        }
        let mut prev = vec![usize::MAX; self.documents.len()];
        let mut q = VecDeque::from([from]);
        prev[from] = from;
        while let Some(cur) = q.pop_front() {
            let mut neighbors: Vec<_> = self.undirected[cur].iter().copied().collect();
            self.sort_doc_indices(&mut neighbors);
            for next in neighbors {
                if prev[next] != usize::MAX {
                    continue;
                }
                prev[next] = cur;
                if next == to {
                    let mut rev = Vec::new();
                    let mut at = to;
                    while at != from {
                        rev.push(at);
                        at = prev[at];
                    }
                    rev.push(from);
                    rev.reverse();
                    return Ok(rev);
                }
                q.push_back(next);
            }
        }
        Err(format!(
            "no path between {:?} and {:?}",
            self.documents[from].name, self.documents[to].name
        ))
    }

    fn largest_component_keys(&self) -> Vec<usize> {
        let mut visited = vec![false; self.documents.len()];
        let mut best: Vec<usize> = Vec::new();
        for idx in 0..self.documents.len() {
            if visited[idx] {
                continue;
            }
            if self.undirected[idx].is_empty() {
                visited[idx] = true;
                continue;
            }
            let mut component = self.collect_component(idx, &mut visited);
            self.sort_doc_indices(&mut component);
            if component.len() > best.len()
                || (component.len() == best.len()
                    && !component.is_empty()
                    && lower(&self.documents[component[0]].name)
                        < lower(&self.documents[best[0]].name))
            {
                best = component;
            }
        }
        best
    }

    pub(crate) fn approx_avg_shortest_path(
        &self,
        sample_count: usize,
        example_count: usize,
        seed: u64,
    ) -> Result<AvgPathReport, String> {
        let component = self.largest_component_keys();
        if component.len() < 2 {
            return Err(
                "largest connected component must contain at least two documents".to_string(),
            );
        }
        if sample_count == 0 {
            return Err("sample count must be positive".to_string());
        }
        let pairs = sample_component_pairs(component.len(), sample_count, seed);
        let samples: Vec<Vec<usize>> = pairs
            .par_iter()
            .map(|&(a, b)| {
                self.shortest_path_indices(component[a], component[b])
                    .unwrap_or_default()
            })
            .collect();
        let total: usize = samples.iter().map(|p| p.len().saturating_sub(1)).sum();
        let avg = total as f64 / samples.len() as f64;
        let mut longer: Vec<_> = samples
            .into_iter()
            .filter(|p| (p.len().saturating_sub(1) as f64) > avg)
            .collect();
        longer.sort_by(|a, b| {
            b.len()
                .cmp(&a.len())
                .then_with(|| path_sort_key(self, a).cmp(&path_sort_key(self, b)))
        });
        longer.truncate(example_count);
        Ok((component.len(), pairs.len(), avg, longer))
    }

    pub(crate) fn all_wanted_pages(&self) -> Vec<WantedPage> {
        #[derive(Default)]
        struct Agg {
            name: String,
            mentions: usize,
            docs: FxHashSet<usize>,
            source_index: FxHashMap<String, usize>,
            sources: Vec<WantedSource>,
        }
        let mut aggs: FxHashMap<String, Agg> = FxHashMap::default();
        for (doc_idx, doc) in self.documents.iter().enumerate() {
            for link in &doc.links {
                if link.resolved.is_some() || link.target_key.is_empty() {
                    continue;
                }
                let mut name = normalize_document_name(&link.display_target);
                if name.is_empty() {
                    name = link.display_target.trim().to_string();
                }
                if name.is_empty() {
                    continue;
                }
                let agg = aggs.entry(link.target_key.clone()).or_insert_with(|| Agg {
                    name,
                    ..Default::default()
                });
                agg.mentions += 1;
                agg.docs.insert(doc_idx);
                let context = if !link.context.trim().is_empty() {
                    link.context.trim().to_string()
                } else if !doc.excerpt.trim().is_empty() {
                    doc.excerpt.trim().to_string()
                } else {
                    "(empty)".to_string()
                };
                let key = format!("{}\0{}", doc.key, context);
                if let Some(&idx) = agg.source_index.get(&key) {
                    agg.sources[idx].mentions += 1;
                } else {
                    agg.source_index.insert(key, agg.sources.len());
                    agg.sources.push(WantedSource {
                        document: doc.name.clone(),
                        context,
                        mentions: 1,
                    });
                }
            }
        }
        let mut pages: Vec<_> = aggs
            .into_values()
            .map(|mut a| {
                a.sources.sort_by(|x, y| {
                    y.mentions
                        .cmp(&x.mentions)
                        .then_with(|| lower(&x.document).cmp(&lower(&y.document)))
                        .then_with(|| lower(&x.context).cmp(&lower(&y.context)))
                });
                WantedPage {
                    name: a.name,
                    mentions: a.mentions,
                    source_documents: a.docs.len(),
                    sources: a.sources,
                }
            })
            .collect();
        pages.sort_by(|a, b| {
            b.mentions
                .cmp(&a.mentions)
                .then(b.source_documents.cmp(&a.source_documents))
                .then_with(|| lower(&a.name).cmp(&lower(&b.name)))
        });
        pages
    }
}

fn sample_component_pairs(
    node_count: usize,
    sample_count: usize,
    seed: u64,
) -> Vec<(usize, usize)> {
    let total_pairs = node_count * (node_count - 1) / 2;
    if sample_count >= total_pairs {
        let mut pairs = Vec::with_capacity(total_pairs);
        for i in 0..node_count {
            for j in i + 1..node_count {
                pairs.push((i, j));
            }
        }
        return pairs;
    }
    let mut rng = StdRng::seed_from_u64(seed);
    let mut seen = FxHashSet::default();
    let mut pairs = Vec::with_capacity(sample_count);
    while pairs.len() < sample_count {
        let mut i = rng.gen_range(0..node_count);
        let mut j = rng.gen_range(0..node_count - 1);
        if j >= i {
            j += 1;
        }
        if i > j {
            std::mem::swap(&mut i, &mut j);
        }
        let key = ((i as u64) << 32) | j as u64;
        if seen.insert(key) {
            pairs.push((i, j));
        }
    }
    pairs
}

fn path_sort_key(vault: &Vault, path: &[usize]) -> String {
    path.iter()
        .map(|&idx| lower(&vault.documents[idx].name))
        .collect::<Vec<_>>()
        .join("\0")
}
